src/data/equiv/algebra.lean

```diff
@@ -324,13 +328,77 @@ instance is_monoid_hom {α β} [monoid α] [monoid β] (h : α ≃* β) : is_mon
 instance is_group_hom {α β} [group α] [group β] (h : α ≃* β) :
   is_group_hom h := { map_mul := h.map_mul }
 
+/-- Two multiplicative isomorphisms agree if they are defined by the
+    same underlying function. -/
+@[extensionality, to_additive
+  "Two additive isomorphisms agree if they are defined by the same underlying function."]
+lemma ext {α β : Type*} [has_mul α] [has_mul β]
+  {f g : mul_equiv α β} (h : ∀ x, f x = g x) : f = g :=
+begin
+  have h₁ := equiv.ext f.to_equiv g.to_equiv h,
+  cases f, cases g, congr,
+  { exact (funext h) },
+  { exact congr_arg equiv.inv_fun h₁ }
+end
+
+attribute [extensionality] add_equiv.ext
+
 end mul_equiv
 
 /-- An additive equivalence of additive groups preserves subtraction. -/
 lemma add_equiv.map_sub {α β} [add_group α] [add_group β] (h : α ≃+ β) (x y : α) :
   h (x - y) = h x - h y :=
 h.to_add_monoid_hom.map_sub x y
 
+
+/-- The group of multiplicative automorphisms. -/
+@[to_additive "The group of additive automorphisms."]
+def mul_aut (α : Type u) [has_mul α] := α ≃* α
+
+namespace mul_aut
+
+variables (α) [has_mul α]
+
+/--
+The group operation on multiplicative automorphisms is defined by
+`λ g h, mul_equiv.trans h g`.
+This means that multiplication agrees with composition, `(g*h)(x) = g (h x)`.
+-/
+instance : group (mul_aut α) :=
+by refine_struct
+{ mul := λ g h, mul_equiv.trans h g,
+  one := mul_equiv.refl α,
+  inv := mul_equiv.symm };
+intros; ext; try { refl }; apply equiv.left_inv
+
+/-- Monoid hom from the group of multiplicative automorphisms to the group of permutations. -/
+def to_perm : mul_aut α →* equiv.perm α :=
+by refine_struct { to_fun := mul_equiv.to_equiv }; intros; refl
+
+end mul_aut
+
+namespace add_aut
+
+variables (α) [has_add α]
+
+/--
+The group operation on additive automorphisms is defined by
+`λ g h, mul_equiv.trans h g`.
+This means that multiplication agrees with composition, `(g*h)(x) = g (h x)`.
+-/
+instance group : group (add_aut α) :=
+by refine_struct
+{ mul := λ g h, add_equiv.trans h g,
+  one := add_equiv.refl α,
+  inv := add_equiv.symm };
+intros; ext; try { refl }; apply equiv.left_inv
```

Comment:
Should there be some `to_additive` line after this definition?
Also, `to_perm` can probably be handled by `to_additive`, so we don't need to duplicate it.

Reply:
to_additive can't handle this, because it would try to replace multiplication in permutations with addition and `monoid_hom` with `add_monoid_hom`.

Reply:
to_additive can't handle this, because it would try to replace multiplication in permutations with addition and `monoid_hom` with `add_monoid_hom`.

Reply:
Sure, but you can still add the connection to the database after manually defining the additive variant, right?

Reply:
I don't think that it's useful, because the types don't match the expectations of `to_additive`. I mean, it will never help `to_additive` to perform the automatic transfer from the multiplicative world the additive world.

Reply:
Fair enough. Let's not bother with it then.